test/test-kq-collision.c

```diff
@@ -0,0 +1,169 @@
+/*
+ * Copyright (c) 2024 Andy Pan <i@andypan.me>
+ *
+ * Redistribution and use in source and binary forms, with or without
+ * modification, are permitted provided that the following conditions
+ * are met:
+ * 1. Redistributions of source code must retain the above copyright
+ *    notice, this list of conditions and the following disclaimer.
+ * 2. Redistributions in binary form must reproduce the above copyright
+ *    notice, this list of conditions and the following disclaimer in the
+ *    documentation and/or other materials provided with the distribution.
+ * 3. The name of the author may not be used to endorse or promote products
+ *    derived from this software without specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE AUTHOR ``AS IS'' AND ANY EXPRESS OR
+ * IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES
+ * OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED.
+ * IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR ANY DIRECT, INDIRECT,
+ * INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT
+ * NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
+ * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
+ * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
+ * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF
+ * THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+ */
+#include "../util-internal.h"
+#include "event2/event-config.h"
+#include "event2/event_compat.h"
+#include "event2/thread.h"
+
+#if defined(EVENT__HAVE_WORKING_KQUEUE)
+
+#include <unistd.h>
+#include <sys/types.h>
+#include <sys/stat.h>
+#ifdef EVENT__HAVE_SYS_TIME_H
+#include <sys/time.h>
+#endif
+#include <fcntl.h>
+#include <stdlib.h>
+#include <stdio.h>
+#include <string.h>
+#include <errno.h>
+#include <pthread.h>
+
+#include <event.h>
+#include <evutil.h>
+
+struct timeval timeout = {3, 0};
+char data[] = "Hello, World!";
+
+#define MAGIC_FD 42 // The old magic number used by kqueue EVFILT_USER
+
+static void
+read_cb(evutil_socket_t fd, short event, void *arg)
+{
+	if (EV_TIMEOUT & event) {
+		printf("%s: Timeout!\n", __func__);
+		exit(1);
+	}
+
+	if (EV_READ & event == 0) {
+		printf("%s: expected EV_READ for pipe but got nothing\n", __func__);
+		exit(1);
+	}
+
+	char buf[16];
+	int n = read(fd, buf, sizeof(buf));
+	if (n == -1) {
+		printf("%s: read error on pipe\n", __func__);
+		exit(1);
+	}
+	buf[n] = '\0';
+	if (strcmp(buf, data) != 0) {
+		printf("%s: read unexpected data from pipe: %s\n", __func__, buf);
+		exit(1);
+	}
+	printf("%s: read the expected data from pipe successfully\n", __func__);
+}
+
+static void*
+trigger_kq(void *arg) {
+	struct event_base *base = arg;
+	event_base_loopcontinue(base);
```

Comment:
Let's add a comment that this is required for notify, and that we can't do it from the same thread, since otherwise notify will not issue any notification

Reply:
Done.